src/main/java/co/thatch/example/metrics/InMemoryMetricsService.java

```diff
@@ -0,0 +1,117 @@
+package co.thatch.example.metrics;
+
+import co.thatch.example.metrics.model.Metric;
+import co.thatch.example.metrics.model.MetricBucket;
+import io.micronaut.runtime.server.event.ServerStartupEvent;
+import io.micronaut.runtime.event.annotation.EventListener;
+import io.micronaut.scheduling.annotation.Async;
+
+import javax.inject.Singleton;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.DoubleAdder;
+import java.util.stream.LongStream;
+
+
+@Singleton
+public class InMemoryMetricsService implements MetricsService {
+    private static final int MAX_WINDOW_SIZE = 1_800;
+    private static final int MAX_WINDOW_SIZE_WITH_BUFFER = MAX_WINDOW_SIZE + 1;
+
+    private final DoubleAdder requestCounter = new DoubleAdder();
+    private final double[] requestCountRecordings = new double[MAX_WINDOW_SIZE_WITH_BUFFER];
+    private final HashMap<Long, MetricHolder> metrics = new HashMap<>();
+    private final PriorityQueue<Long> timestamps = new PriorityQueue<>();
+
+    @Async
+    @EventListener
+    public void onStartup(ServerStartupEvent event) throws InterruptedException {
+        while (true) {
+            requestCountRecordings[Math.toIntExact(Instant.now().getEpochSecond() % requestCountRecordings.length)] = requestCounter.doubleValue();
+            Thread.sleep(50);
+        }
+    }
+
+    @Override
+    public void incrementRequestCount(String key, Instant requestInstant) {
+//        CompletableFuture.runAsync(() -> {
+            requestCounter.add(1.0);
+            Long timestamp = requestInstant.getEpochSecond();
+            synchronized (this) {
+                MetricHolder metricsForTimestamp = metrics.get(timestamp);
+                if (metricsForTimestamp != null) {
+                    metricsForTimestamp.incrementRequestCountForKey(key);
+                } else {
+                    metrics.put(timestamp, new MetricHolder(key));
+                    timestamps.add(timestamp);
+                    if (timestamps.peek() < timestamp - 30) {
+                        metrics.remove(timestamps.poll());
+                    }
+                }
+            }
+//        });
+    }
+
+    @Override
+    public double getQPS(Duration window) {
+        double startCount = requestCountRecordings[(Math.toIntExact(Instant.now().minus(window).minusSeconds(1).getEpochSecond()) % requestCountRecordings.length)];
```

Comment:
It feels like this solution is sticking quite closely to the prometheus model we discussed? I think that if you move to a model that doesn't assume a single monotonic counter, you could combine both the average QPS and the histogram into a combined approach.